app/uk/gov/hmrc/helptosave/connectors/HelpToSaveProxyConnector.scala

```diff
@@ -159,9 +163,61 @@ class HelpToSaveProxyConnectorImpl @Inject() (http:              WSHttp,
     )
   }
 
+  override def getTransactions(
+      nino: String, systemId: String, correlationId: String)(
+      implicit
+      hc: HeaderCarrier, ec: ExecutionContext): Result[Option[Transactions]] = {
+
+    val url = s"$proxyURL/help-to-save-proxy/nsi-services/transactions" ? ("nino" -> nino) & ("correlationId" -> correlationId) & ("version" -> getTransactionsVersion) & ("systemId" -> systemId)
+
+    val timerContext = metrics.getTransactionsTimer.time()
+    EitherT[Future, String, Option[Transactions]](
+      http.get(url).map[Either[String, Option[Transactions]]] {
+        response ⇒
+          val _ = timerContext.stop()
+          response.status match {
+            case Status.OK ⇒
+              val result = response.parseJsonWithoutLoggingBody[NsiTransactions].flatMap(nsiTransactions ⇒
+                Transactions(nsiTransactions).toEither.bimap(s ⇒ s"[${s.toList.mkString(",")}]", Some(_)))
+
+              result.bimap(
```

Comment:
I tried to change this to bimap once but got an unwanted behaviour change (lack of transformation of the error message when `parseJsonWithoutLoggingBody` fails. See `NGC-3251-combine-bimaps` branch - `HelpToSaveProxyConnectorSpec` fails in that branch.
Perhaps we can have a look at the code together?

Reply:
the problem might be that `parseJsonWithoutLoggingBody` returns `Either[String,NsiTransactions]` but `Transactions(nsiTransactions).toEither` returns   `Either[NonEmptyList[String],Transactions]`. The left types don't match up which means the flatmapping/bimapping might not work as you expect

Reply:
I managed to get the types to line up (i.e. write something that compiles), the problem is that the left-hand function in the bimap is not called if `parseJsonWithoutLoggingBody` returns a `Left` because `Either.flatMap` does not invoke the passed function on a `Left` only on a `Right`.

If there was such a thing as `Either.flatBimap` then we could use that but as it stands I can't think of a good way to avoid having 2 bimaps.

Reply:
I've got it working now by using `.left.map` - see https://github.com/hmrc/help-to-save/blob/NGC-3251-combine-bimaps/app/uk/gov/hmrc/helptosave/connectors/HelpToSaveProxyConnector.scala#L181

Let me know if you prefer that version, if so I'll apply that change to this branch and also do the same change in `getAccount`

Reply:
I see what you mean now. The problem is the `bimap` is happening in the `flatMap`. The cleanest solution would probably be then to do the `flatMap` which builds up the result using the different functions used to build it up and then `bimap` on that result to handle the errors more consistently. Something like:
```scala
val result = for {
                nsiTransactions ← response.parseJsonWithoutLoggingBody[NsiTransactions].leftMap(NonEmptyList.one)
                transactions ← Transactions(nsiTransactions).toEither
              } yield {
                transactions
              }

              result.bimap(
                {
                  errors ⇒
                    metrics.getTransactionsErrorCounter.inc()
                    pagerDutyAlerting.alert("Could not parse get transactions response")
                    s"""Could not parse transactions response from NS&I, received 200 (OK), error=[${errors.toList.mkString(",")}]"""
                }, { transactions ⇒
                  logger.info("Call to get transactions successful", nino, "correlationId" → correlationId)
                  Some(transactions)
                }
              )
```
That's almost what you've done in  the `.left.map` implementation, but with the `for-comprehension` it's a bit easier to follow

Reply:
Thanks, I've used your suggested implementation. There is a [slight change in the error string that ist returned in some circumstances](https://github.com/hmrc/help-to-save/pull/140/commits/ceeee8dae4dda12aefff3c5508ab2e9fc5c22347#diff-637ff59672f383b2286acaf4344f6b8bR495) but I think that's fine or even an improvement.